### PR TITLE
Adds method to user model for creation and some tests to check it

### PR DESCRIPTION
**Pull request by christiangelone**
## Summary

 * Adds functionality for creation of user, in a backend perspective
 * Tests creation of users for creation of use

## Things to consider in the future:
  * Creation of users should be db / framework agnostic 

## Trello Card

https://trello.com/c/U88apt6L

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+'use strict';
+
+const bcrypt = require('bcryptjs');
+const logger = require('../logger');
+const User = require('../models').User;
+
+exports.create = (req, res, next) => {
+  const userObj = req.body || {};
+  User.isPwdLengthGreater(userObj, 8)
```
Comment (gonziesc) [inline app/controllers/users.js]: you have to return the promise

File: app/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    .then(() => {
+      bcrypt.hash(userObj.password, 10).then(hashedPwd => {
```
Comment (gonziesc) [inline app/controllers/users.js]: you dont need to wait for the user.ispwdfLengthGreater for calling this one, you can do a promise.all with both

File: app/controllers/users.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        userObj.password = hashedPwd;
+        User.createModel(userObj)
```
Comment (gonziesc) [inline app/controllers/users.js]: return promise. also, i would call it just create

File: app/controllers/users.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          .then(createdUser => {
+            logger.log({
+              level: 'info',
+              message: createdUser.firstname
+            });
+            res.status(201).send();
+          })
+          .catch(err => {
+            if (err.name === 'SequelizeValidationError') {
```
Comment (gonziesc) [inline app/controllers/users.js]: you can use the error manager for this

File: app/controllers/users.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            logger.log({
+              level: 'error',
+              message: JSON.stringify(err.errors, null, 2)
+            });
+            res.status(500).send();
+          });
+      });
+    })
+    .catch(() => {
+      const someErrors = [
```
Comment (gonziesc) [inline app/controllers/users.js]: same here

File: app/controllers/users.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        {
+          error: 'password should be greater than or equal to 8 characters'
```
Comment (gonziesc) [inline app/controllers/users.js]: abstract the message to other part of the code, like the error manager.

File: app/models/index.js [hunk 7]
```diff
@@ -8,6 +8,10 @@ const fs = require('fs'),
 
 const sequelize = new Sequelize(dbConfig.database, dbConfig.username, dbConfig.password, dbConfig);
 
+sequelize.authenticate().then(() => {
```
Comment (gonziesc) [inline app/models/index.js]: delete this

File: app/models/user.js [hunk 8]
```diff
@@ -5,15 +5,58 @@ module.exports = (sequelize, DataTypes) => {
...
+          notEmpty: false
+        }
+      },
+      email: {
+        type: DataTypes.STRING,
+        allowNull: false,
+        required: true,
+        unique: true,
+        validate: {
+          isEmail: { args: true, msg: 'not a valid email' },
```
Comment (gonziesc) [inline app/models/user.js]: lets validate this before going to the database

File: app/models/user.js [hunk 9]
```diff
@@ -5,15 +5,58 @@ module.exports = (sequelize, DataTypes) => {
...
+          notEmpty: false,
+          len: { args: [0, 100], msg: 'email cant be bigger than 100 characters' }
```
Comment (gonziesc) [inline app/models/user.js]: same this

File: app/models/user.js [hunk 10]
```diff
@@ -5,15 +5,58 @@ module.exports = (sequelize, DataTypes) => {
...
+      }
     },
     {}
   );
-  User.associate = function(models) {
-    // associations can be defined here
+  User.associate = models => {};
+
+  User.isPwdLengthGreater = (user, length) => {
+    return new Promise((res, rej) => {
```
Comment (gonziesc) [inline app/models/user.js]: this is synchronous, there is no need to make a promise for this

File: app/models/user.js [hunk 11]
```diff
@@ -5,15 +5,58 @@ module.exports = (sequelize, DataTypes) => {
...
+      if (user.password && user.password.length < length) {
+        rej();
+      } else {
+        res();
+      }
+    });
+  };
+
+  User.createModel = user => {
```
Comment (gonziesc) [inline app/models/user.js]: this method is doing nothing

File: test/controllers/user.test.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    email: 'john.doe@wolox.com.ar',
+    password: 'johndoereloaded'
+  };
+
+  it('Should create a user', done => {
+    request
+      .post('/users')
+      .send(validUser)
+      .then(res => {
+        res.should.have.status(201);
```
Comment (gonziesc) [inline test/controllers/user.test.js:24]: validate that the user is created

File: test/controllers/user.test.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  it('Should not create a user without email', done => {
+    request
+      .post('/users')
+      .send(userWithoutEmail)
+      .then(res => {
+        done(new Error('This should not be called'));
+      })
+      .catch(err => {
+        err.should.have.status(400);
+        err.response.should.be.json;
```
Comment (gonziesc) [inline test/controllers/user.test.js]: validate message

File: test/controllers/user.test.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  it('Should not create an user with a short password', done => {
+    request
+      .post('/users')
+      .send(userWithShortPassword)
+      .then(res => {
+        done(new Error('This should not be called'));
+      })
+      .catch(err => {
+        err.should.have.status(400);
+        err.response.should.be.json;
```
Comment (gonziesc) [inline test/controllers/user.test.js]: same

File: test/controllers/user.test.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+  it('Should not create an empty user', done => {
+    request
+      .post('/users')
+      .send(emptyUser)
+      .then(res => {
+        done(new Error('This should not be called'));
+      })
+      .catch(err => {
+        err.should.have.status(400);
+        err.response.should.be.json;
```
Comment (gonziesc) [inline test/controllers/user.test.js]: same

File: test/services/user.test.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
+const assert = require('chai').assert;
```
Comment (gonziesc) [inline test/services/user.test.js]: delete this file

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/controllers/users.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+const bcrypt = require('bcryptjs');
+const logger = require('../logger');
+const errors = require('../errors');
+
+const User = require('../models').User;
+
+exports.create = (req, res, next) => {
+  const userObj = req.body || {};
+  bcrypt
```
Comment (gonziesc) [inline app/controllers/users.js]: return promise

File: app/controllers/users.js [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    .hash(userObj.password, 10)
+    .then(hashedPwd => {
+      userObj.password = hashedPwd;
+      User.createModel(userObj)
```
Comment (gonziesc) [inline app/controllers/users.js]: return promise

File: app/models/user.js [hunk 19]
```diff
@@ -5,15 +5,33 @@ module.exports = (sequelize, DataTypes) => {
...
+        allowNull: false,
+        required: true
+      }
     },
     {}
   );
-  User.associate = function(models) {
-    // associations can be defined here
+
+  User.createModel = user => {
```
Comment (gonziesc) [inline app/models/user.js]: this is useless

File: app/routes.js [hunk 20]
```diff
@@ -1,7 +1,18 @@
...
+const validate = new Validator({ allErrors: true }).validate;
 
 exports.init = app => {
-  // app.get('/endpoint/get/path', [], controller.methodGET);
-  // app.put('/endpoint/put/path', [], controller.methodPUT);
-  // app.post('/endpoint/post/path', [], controller.methodPOST);
+  app.post('/users', [validate({ body: UserJsonSchema.forCreate })], UsersController.create);
+
+  // Handles body validation errors
+  app.use((err, req, res, next) => {
```
Comment (gonziesc) [inline app/routes.js]: do you need this?

File: test/controllers/user.test.js (hunk 12 above)
Comment (gonziesc) [reply test/controllers/user.test.js:24]: check it in the database

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/routes.js (hunk 20 above)
Comment (christiangelone) [reply app/routes.js]: Yes to process any json schema validation issue

**Comment by christiangelone** [comment]
All done!

File: app/controllers/users.js (hunk 17 above)
Comment (gonziesc) [reply app/controllers/users.js]: again

File: test/controllers/user.test.js (hunk 12 above)
Comment (gonziesc) [reply test/controllers/user.test.js:24]: same

File: test/models/user.test.js [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+    lastname: 'Doe',
+    email: 'john.doe@wolox.com.ar',
+    password: 'johndoereloaded'
+  };
+
+  it('Should create a user', done => {
+    User.create(validUser)
+      .then(createdUser => {
+        createdUser.should.have.property('id');
+        User.findById(createdUser.id)
```
Comment (gonziesc) [inline test/models/user.test.js]: move this validation to the create endpoint, and delete this file

**Review by gonziesc — CHANGES_REQUESTED** [review]

**Comment by gonziesc** [comment]
Lasts comments!

File: app/controllers/users.js (hunk 17 above)
Comment (gonziesc) [reply app/controllers/users.js]: return the bcrypt promise

Comment (gonziesc) [reply app/controllers/users.js]: return the bcrypt promise

File: app/errors.js [hunk 22]
```diff
@@ -5,3 +5,20 @@ const internalError = (message, internalCode) => ({
...
+        return {
+          error: err.message
+        };
+      })
+    },
+    exports.DATABASE_ERROR
+  );
+
+exports.NOT_FOUND = 'not_found';
+exports.notFound = resource => internalError(`${resource} not found`, exports.NOT_FOUND);
```
Comment (gonziesc) [inline app/errors.js:24]: what is this doing here?

File: test/controllers/user.test.js (hunk 12 above)
Comment (gonziesc) [reply test/controllers/user.test.js:24]: the idea was that you validate here with the db, not with another endpoint

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/controllers/users.js [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+      next(errors.databaseError(err));
+    })
+    .catch(err => {
+      next(errors.defaultError(err));
+    });
+};
+
+exports.get = (req, res, next) => {
+  const anId = req.params.id;
+  User.find({
```
Comment (gonziesc) [inline app/controllers/users.js]: return

File: app/controllers/users.js [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+      logger.log({ level: 'error', message: JSON.stringify(err, null, 2) });
+      next(errors.databaseError(err));
+    })
+    .catch(err => {
+      next(errors.defaultError(err));
+    });
+};
+
+exports.get = (req, res, next) => {
+  const anId = req.params.id;
```
Comment (gonziesc) [inline app/controllers/users.js]: just id for name

**Review by gonziesc — CHANGES_REQUESTED** [review]

**Review by gonziesc — APPROVED** [review]

**Pull request merged**